### ISSUE TITLE
Leex generates code that causes warnings from the compiler about not being able to match clauses.

### ISSUE THREAD

**Issue by skynet1** [erlang/otp#4918]
**Describe the bug**
Leex generates code that compiles with the following warnings:
`src/jsx.erl:694: Warning: this clause cannot match because a previous clause at line 692 always matches
src/jsx.erl:728: Warning: this clause cannot match because a previous clause at line 726 always matches
src/jsx.erl:928: Warning: this clause cannot match because a previous clause at line 926 always matches
src/jsx.erl:1218: Warning: this clause cannot match because a previous clause at line 1216 always matches
src/jsx.erl:1526: Warning: this clause cannot match because a previous clause at line 1524 always matches
src/jsx.erl:1874: Warning: this clause cannot match because a previous clause at line 1872 always matches
`

**To Reproduce**
Run 
`leex:file("jsx.xrl"). c("jsx.erl").`

**Expected behavior**
The code generated shouldn't generate code that has redundant clauses.

**Affected versions**
Erlang V11.2.1

**Additional context**
Here's the code:
```
Definitions.
Rules.
"(\\.|[^\\"])*"	: {token, {string_lit, TokenChars}}.
/(\\.|[^\\/]|\[(/|\\.|[^\\/])+\]|\((\\.|/|[^\\/])+\)+)*/[g]?	: {token, {regex_lit, TokenChars}}. %% The regex must not contain a / unless it is escaped, but it can contain an unescaped / if it's inside a class, or in unescaped paranthesis.

'(\\.|[^'\\])*'	: {token, {string_lit, TokenChars}}.
//.*\n	: skip_token.
/\*([^*][\r\n]|(\*+([^*/]|[\r\n])))*\*+/	: skip_token. %% Got this from here: https://blog.ostermiller.org/finding-comments-in-source-code-using-regular-expressions/
[_$a-zA-z][_$a-zA-Z0-9]*	: {token, {'identifier', TokenChars}}.

Erlang code.

exponent(0) -> 1;
exponent(X) -> 10 * exponent(X-1).

fixed_to_integer(L) ->
	FH = get_fixed_first_half(L),
	SH = get_fixed_second_half(L),
	FH1 = list_to_integer(FH),
	SH1 = list_to_integer(SH),
	FH1 * exponent(SH1).

get_fixed_first_half([]) -> "";
get_fixed_first_half([$e|_]) -> "";
get_fixed_first_half([$E|_]) -> "";
get_fixed_first_half([X|Rest]) -> [X|get_fixed_first_half(Rest)].

get_fixed_second_half([]) -> [];
get_fixed_second_half([$e|Rest]) -> Rest;
get_fixed_second_half([$E|Rest]) -> Rest;
get_fixed_second_half([_|Rest]) -> get_fixed_second_half(Rest).
```

**Comment by uabboli** [comment]
I'm not familiar with the Leex implementation, but a patch that seems to work is this:

```diff --git a/lib/parsetools/src/leex.erl b/lib/parsetools/src/leex.erl
index 90d1b9f71e..37f3fb7114 100644
--- a/lib/parsetools/src/leex.erl
+++ b/lib/parsetools/src/leex.erl
@@ -1547,7 +1547,10 @@ pack_trans(Trs) -> pack_trans(Trs, []).
 %%     Trs1.
 
 pack_trans([{{C,C},S}|Trs], Pt) ->         % Singletons to the head
-    pack_trans(Trs, [{C,S}|Pt]);
+    case lists:member({C,S}, Pt) of
+        true -> pack_trans(Trs, Pt);
+        false -> pack_trans(Trs, [{C,S}|Pt])
+    end;
 %% Special detection and handling of $\n.
 pack_trans([{{Cf,$\n},S}|Trs], Pt) ->
     pack_trans([{{Cf,$\n-1},S}|Trs], [{$\n,S}|Pt]);
```

I wonder if @rvirding has a better suggestion.

**Comment by uabboli** [comment]
Merged commit 7e875cce407e1ac68ef3e66051bc6a3164980fdf into erlang:maint.
